File: 5-Problem-A-Day/50. Pow(x, n).java
class Solution {
    public double myPow(double x, int n) {
        if (n == Integer.MIN_VALUE) {
            double half = halfPow(1 / x, -n/2);
            return half * half;
        }            
        if (n < 0) {
            n = -n;
            x = 1 / x;
        }
        return halfPow(x, n);
    }
    
    private double halfPow(double x, int n) {
        if (n == 0)
            return 1.0;
        double half = halfPow(x, n /2);
        if (n % 2 == 0)
            return half * half;
        else
            return half * half * x;
    }        
}

File: 2024/50. Pow(x, n).cpp
class Solution {
public:
    double myPow(double x, int n) {
        double pos = posPow(x, abs(n));
        return n > 0 ? pos : 1 / pos;
    }
    double posPow(double x, int n ) {
        if (n == 0) {
            return 1;
        }
        double half = myPow(x, n / 2);
        if (n % 2 == 1) {
            return half * half * x;
        } else {
            return half * half;
        }
    }
};